addons/web_editor/static/src/js/wysiwyg/widgets/media.js

```diff
@@ -454,52 +470,76 @@ var FileWidget = SearchableMediaWidget.extend({
-                    self.media = self.$media[0];
-                }
-                var href = '/web/content/' + img.id + '?';
-                if (!img.public && img.access_token) {
-                    href += _.str.sprintf('access_token=%s&', img.access_token);
-                }
-                href += 'unique=' + img.checksum + '&download=true';
-                self.$media.attr('href', href);
-                self.$media.addClass('o_image').attr('title', img.name).attr('data-mimetype', img.mimetype);
+        if (img.image_src) {
+            var src = img.image_src;
+            if (!img.public && img.access_token) {
+                src += _.str.sprintf('?access_token=%s', img.access_token);
             }
+            if (!this.$media.is('img')) {
 
-            self.$media.attr('alt', img.alt);
-            var style = self.style;
-            if (style) {
-                self.$media.css(style);
+                // Note: by default the images receive the bootstrap opt-in
+                // img-fluid class. We cannot make them all responsive
+                // by design because of libraries and client databases img.
+                this.$media = $('<img/>', {class: 'img-fluid o_we_custom_image'});
+                this.media = this.$media[0];
             }
-
-            // Remove crop related attributes
-            if (self.$media.attr('data-aspect-ratio')) {
-                var attrs = ['aspect-ratio', 'x', 'y', 'width', 'height', 'rotate', 'scale-x', 'scale-y', 'crop:originalSrc'];
-                self.$media.removeClass('o_cropped_img_to_save');
-                _.each(attrs, function (attr) {
-                    self.$media.removeData(attr);
-                    self.$media.removeAttr('data-' + attr);
-                });
+            src = new URL(src, window.location.origin);
+            src.searchParams.set('cachebuster', Date.now());
+            this.$media.attr('src', src.href);
+            this.$media[0].dataset.removeCacheBuster = 'true';
+        } else {
+            if (!this.$media.is('a')) {
+                $('.note-control-selection').hide();
+                this.$media = $('<a/>');
+                this.media = this.$media[0];
+            }
+            var href = '/web/content/' + img.id + '?';
+            if (!img.public && img.access_token) {
+                href += _.str.sprintf('access_token=%s&', img.access_token);
             }
-            return self.media;
+            href += 'unique=' + img.checksum + '&download=true';
+            this.$media.attr('href', href);
+            this.$media.addClass('o_image').attr('title', img.name).attr('data-mimetype', img.mimetype);
+        }
+        const media = this.$media[0];
+        if (img.autoOptimized) {
+            const removeIfUnused = JSON.parse(media.dataset.removeIfUnused || '[]');
+            media.dataset.removeIfUnused = JSON.stringify(removeIfUnused.concat([img.id]));
+        }
+        media.dataset.currentlyUsed = img.id;
```

Comment:
I don't understand, if the user removes the image, don't we lose a reference to the images to remove? Can't you just trigger_up an event towards the object that is in charge of the garbage collect to notify him the ids he will have to check? That way, no DOM involved, no JSON parsing, etc. The same way, is it necessary to use `currentlyUsed`? What about just check if any /web/image/XXX is present for each XXX you have to potentially remove?

Reply:
I'll see about the first part, but the second part won't work with unsplash attachments. Although now, unsplash attachments also have the attachment id in their URL so I could check for both url formats, but that seems a little fragile (if the url format ever changes, it will break), and I'll also have to check for every image in the editable DOM and not just those that have actually been changed.

Reply:
Ok, then only the first part :+1: